sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,641 @@
+            }
+
+            return stdout;
+        }
+
+        /// <summary>
+        /// Runs a CLI command in a child process, returning a task for the process's exit. Both stdout
+        /// and stderr are redirected to process.stdout and process.stder by default.
+        /// </summary>
+        /// <param name="cmd">Command name to run.</param>
+        /// <param name="args">An array of command arguments.</param>
+        /// <param name="logResource">Used to specify the resource to associate command output with. Stderr messages
+        /// are always sent (since they may contain important information about something that's gone wrong).
+        /// Stdout messages will be logged ephemerally to this resource.  This lets the user know there is
+        /// progress, without having that dumped on them at the end.  If an error occurs though, the stdout
+        /// content will be printed.</param>
+        /// <param name="reportFullCommandLine"></param>
+        /// <param name="reportErrorAsWarning"></param>
+        /// <param name="stdin">If defined, its contents are piped into stdin for the child process.</param>
+        /// <param name="env"></param>
+        /// <returns></returns>
+        private static async Task<(int, string)> RunCommandThatCanFail(
+            string cmd,
+            string[] args,
+            Resource logResource,
+            bool reportFullCommandLine,
+            bool reportErrorAsWarning,
+            string? stdin = null,
+            ImmutableDictionary<string, string>? env = null)
+        {
+            // Let the user ephemerally know the command we're going to execute.
+            Log.Info($"Executing {GetCommandLineMessage(cmd, args, reportFullCommandLine, env)}", logResource, ephemeral: true);
+
+            var process = new Process();
+            process.StartInfo.FileName = cmd;
+            process.StartInfo.Arguments = Utils.EscapeArguments(args);
+            process.StartInfo.RedirectStandardInput = stdin != null;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.RedirectStandardError = true;
+
+            try
+            {
+                process.Start();
+
+                if (stdin != null)
+                {
+                    process.StandardInput.Write(stdin);
+                    process.StandardInput.Close();
+                }
+
+                var stdout = await process.StandardOutput.ReadToEndAsync().ConfigureAwait(false);
+
+                // We can't stream these stderr messages as we receive them because we don't knows at
+                // this point because Docker uses stderr for both errors and warnings.  So, instead, we
+                // just collect the messages, and wait for the process to end to decide how to report
+                // them.
+                var stderr = await process.StandardError.ReadToEndAsync().ConfigureAwait(false);
```

Comment:
i'm always terrified of stdout/err in .net.  is there any chance the async calls could hang because the first one is awaiting completion, but the process isn't completing because it's trying to write to the other, and nothing is reading from it?

In other words, should we kick off two Task.Runs to ensure that the readings will happen concurrently so we don't get any deadlocks?

Reply:
that's a valid point... I guess we should do whatever we can to prevent deadlocks

Reply:
I realized that the node SDK was broadcasting output lines as they come, while my .NET implementation was waiting for the complete output. Now, I changed to async output consumption based on `OutputDataReceived` event. This should also address the potential deadlock.

Reply:
should we not do this for streams?  i.e. why do you still ReadToEndAsync with .StandardError?

Reply:
Changed. Are we good to go?